### PR TITLE
Fix testing suite name conflict for workflow and activity (#887)

### PR DESCRIPTION
**Pull request by itechdima**
## What was changed

Splitted activity and workflow mocks so you're able to use same names to mock workflows and activities, but the logic of all exported functions kept the same

## Why?

To be able to use same workflow and activity function names when they are struct methods (avoid tautology, etc.)

## Checklist

1. Closes 
#887 

2. How was this tested:

Added `Test_SameWorkflowAndActivityNames` function that uses same workflow and activity name

3. Any docs updates needed?

No

### CODE REVIEW

File: internal/workflow_testsuite.go [hunk 1]
```diff
@@ -942,16 +951,28 @@ func (e *TestWorkflowEnvironment) AssertExpectations(t mock.TestingT) bool {
...
+	}
+	return true
 }
 
 // AssertNotCalled asserts that the method was not called with the given arguments.
 // See AssertCalled for more info.
 func (e *TestWorkflowEnvironment) AssertNotCalled(t mock.TestingT, methodName string, arguments ...interface{}) bool {
-	return e.mock.AssertNotCalled(t, methodName, arguments...)
+	dummyT := &testing.T{}
+	if !(e.workflowMock.AssertNotCalled(dummyT, methodName, arguments...) || e.activityMock.AssertNotCalled(dummyT, methodName, arguments...)) {
```
Comment (Quinn-With-Two-Ns) [inline internal/workflow_testsuite.go:965]: Did you mean to check `activityMock` here?

File: internal/workflow_testsuite.go [hunk 2]
```diff
@@ -942,16 +951,28 @@ func (e *TestWorkflowEnvironment) AssertExpectations(t mock.TestingT) bool {
...
+		return e.workflowMock.AssertNotCalled(t, methodName, arguments...) && e.activityMock.AssertNotCalled(t, methodName, arguments...)
+	}
+	return true
 }
 
 // AssertNumberOfCalls asserts that a method was called expectedCalls times.
 func (e *TestWorkflowEnvironment) AssertNumberOfCalls(t mock.TestingT, methodName string, expectedCalls int) bool {
-	return e.mock.AssertNumberOfCalls(t, methodName, expectedCalls)
+	dummyT := &testing.T{}
+	if !(e.workflowMock.AssertNumberOfCalls(dummyT, methodName, expectedCalls) || e.activityMock.AssertNumberOfCalls(dummyT, methodName, expectedCalls)) {
```
Comment (Quinn-With-Two-Ns) [inline internal/workflow_testsuite.go:974]: Did you mean to check `activityMock` here?

File: internal/workflow_testsuite.go [hunk 3]
```diff
@@ -942,16 +951,28 @@ func (e *TestWorkflowEnvironment) AssertExpectations(t mock.TestingT) bool {
...
+	dummyT := &testing.T{}
+	if !(e.workflowMock.AssertCalled(dummyT, methodName, arguments...) || e.activityMock.AssertCalled(dummyT, methodName, arguments...)) {
+		return e.workflowMock.AssertCalled(t, methodName, arguments...) && e.activityMock.AssertCalled(t, methodName, arguments...)
+	}
+	return true
 }
 
 // AssertNotCalled asserts that the method was not called with the given arguments.
 // See AssertCalled for more info.
 func (e *TestWorkflowEnvironment) AssertNotCalled(t mock.TestingT, methodName string, arguments ...interface{}) bool {
```
Comment (Quinn-With-Two-Ns) [inline internal/workflow_testsuite.go:963]: I think this fix is acceptable. I don't believe it is backwards compatible technically but if fixing this breaks their test the test was probably wrong to begin with.

Do you think we need workflow and activity versions of this and `AssertNumberOfCalls`  that differentiate between activity and workflow mocks?

Comment (Quinn-With-Two-Ns) [reply internal/workflow_testsuite.go:963]: also  `AssertNumberOfCalls`

Comment (Quinn-With-Two-Ns) [reply internal/workflow_testsuite.go:963]: @cretz in the issue you mention backwards compatibility concerns. Did you have other concerns then these 3 functions?

Comment (itechdima) [reply internal/workflow_testsuite.go:963]: Do you mean `AssertActivityCalled`/`AssertWorkflowCalled` and others? 
We might add them as well with keeping general AssertCalled (and others)

Comment (cretz) [reply internal/workflow_testsuite.go:963]: > Did you have other concerns then these 3 functions?

That was my only concern I can recall, our high-level assertion calls. I just want to make sure we don't break anyone's existing testing code.

> Do you mean AssertActivityCalled/AssertWorkflowCalled and others?
> We might add them as well with keeping general AssertCalled (and others)

Hrmm, I don't have a strong opinion here. I am guessing we don't want to give `mock.Mock` to the user, so I guess we might as well expose.

Comment (itechdima) [reply internal/workflow_testsuite.go:975]: > I just want to make sure we don't break anyone's existing testing 

I don't know the best approach here, but I can agree with @Quinn-With-Two-Ns that if these changes break anyones test, the test itself was logically wrong

Another point is that temporal allows you to use same names for activity and workflows, but SDK doesn't allow you to test it :)

Comment (Quinn-With-Two-Ns) [reply internal/workflow_testsuite.go:975]: @cretz to be clear I think this change could cause some tests that passed before to fail, but those tests depended on workflow and activity mocks overriding each other which is a bug. I am also not sure how popular these functions even are. I think it is worth the risk of breaking some bad tests to fix this weird footgun for users.

Comment (cretz) [reply internal/workflow_testsuite.go:975]: Yeah, I definitely don't mind if someone was relying on the name ambiguity and they break. But other than that, if we feel proper tests won't break, I'm good.

**Comment by Quinn-With-Two-Ns** [comment]
@itechdima I am seeing very consistent CI failures do you know what causes them?

**Comment by itechdima** [comment]
@Quinn-With-Two-Ns 
Hope that I fixed them :)

**Comment by itechdima** [comment]
I'm not sure why another test has failed..
could it be flaky test?

**Comment by cretz** [comment]
Rerunning to see

**Comment by itechdima** [comment]
@cretz 
Thanks :) 
Looks like there were some flakiness

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Pull request merged**